File: 1114. Print in Order/Solution.java
class Foo {
    Semaphore firstDone = new Semaphore(0);
    Semaphore secondDone = new Semaphore(0);

    public Foo() {}

    public void first(Runnable printFirst) throws InterruptedException {
        // printFirst.run() outputs "first". Do not change or remove this line.
        printFirst.run();
        firstDone.release();
    }

    public void second(Runnable printSecond) throws InterruptedException {
        firstDone.acquire();
        // printSecond.run() outputs "second". Do not change or remove this line.
        printSecond.run();
        secondDone.release();
    }

    public void third(Runnable printThird) throws InterruptedException {
        secondDone.acquire();
        // printThird.run() outputs "third". Do not change or remove this line.
        printThird.run();
    }
}

File: 1114. Print in Order/Solution.py
import threading
from typing import Callable


class Foo:
    def __init__(self):
        self.firstDone = threading.Lock()
        self.firstDone.acquire()
        self.secondDone = threading.Lock()
        self.secondDone.acquire()

    def first(self, printFirst: Callable[[], None]) -> None:
        # printFirst() outputs "first". Do not change or remove this line.
        printFirst()
        self.firstDone.release()

    def second(self, printSecond: Callable[[], None]) -> None:
        with self.firstDone:
            # printSecond() outputs "second". Do not change or remove this line.
            printSecond()
            self.secondDone.release()

    def third(self, printThird: Callable[[], None]) -> None:
        with self.secondDone:
            # printThird() outputs "third". Do not change or remove this line.
            printThird()
